### DISCUSSION TITLE
xterm.js / portable_pty duplicate lines when resizing

### DISCUSSION THREAD

**Discussion by BrendanMartin** [wezterm/wezterm#3363] [Q&A]
Xterm.js (using xterm fit addon) sends the new rows and cols to portable_pty on resize, but I found that I get duplicate prompt lines when resizing drastically.

https://user-images.githubusercontent.com/8090350/227723739-e18c09c3-23d2-4f69-bff9-76ef1dc8837e.mp4

Using a 200ms timeout from JS for sending resize events seems to fix it, but I'm wondering if that's the "correct" approach. VS Code uses xterm.js and node-pty and their resizing is really smooth.

**Comment by wez** [comment]
I don't understand what I'm looking at; that isn't wezterm. As far as I understand, xterm.js doesn't use any of wezterm's components.  What exactly is that, and what exactly are you asking?

**Reply by BrendanMartin** [reply]
I'm sorry if this is the wrong place to talk about this. I saw the portable_pty crate was part of wezterm and wondered if anyone could explain why this is happening. You're looking at a side project I'm working on that uses portable_pty on the back end and xterm.js on the front end.

**Reply by wez** [reply]
I see. Resize behavior isn't something that portable_pty directly influences: it's just a interface to the kernel pty layer that your application can use to signal resizes and consume the output data stream.

The extra prompt redraws are because the application on the other end of the pty detected the resize event and repainted the screen.  There's always the possibility of a resize racing with a repaint, and every terminal has glitches in some situations with this.

The only way to completely avoid it is to fully de-bounce the resize event and only pass it on the pty when the size has been stable for a while, but that doesn't look as smooth as live window resizing.

For an additional complication: ConPTY (the windows PTY layer), can sometimes "editorialize" the data stream and synthesize a full screen repaint in some surprising situations, so I'd recommend that you test on both Windows and unix platforms before you settle on your resize behavior.

**Reply by BrendanMartin** [reply]
Thanks for the detailed response.

"The extra prompt redraws are because the application on the other end of the pty detected the resize event and repainted the screen"
By the application on the other end of pty, do you mean xterm.js in my case?

**Reply by wez** [reply]
the shell, or whatever program was spawned into the pty
